pkgs/native_toolchain_c/test/cbuilder/cbuilder_build_failure_test.dart

```diff
@@ -30,20 +30,30 @@ void main() {
     await File.fromUri(addCUri).writeAsString(addCBrokenContents);
     const name = 'add';
 
-    final buildConfig = BuildConfig.build(
-      supportedAssetTypes: [CodeAsset.type],
+    final buildConfigBuilder = BuildConfigBuilder()
+      ..setupHookConfig(
```

Comment:
This code is much less nice than what it was before. Ditto for all the other tests.

I guess that's just what it is, much less nice API for our users due to having to deal with extensibility.

Reply:
Then we very much disagree here:

* *before* we have a whole bunch of parameters being passed down, some of them being related to others (e.g. `cCompilerConfig`, `linkModePreference`) others completely unrelated (e.g. `dryRun`, `packageName`)
* *now* we group the things together that belong together

Imagine we had to extend the parameters here with fonts, icons, proguard rules, ... we'd end up in a long list of unrelated things mashed together into a long parameter list.

How is the new way not nicer?

Reply:
* You can easily forget to call `setupBuildConfig`. The resulting `BuildConfig` no longer correct by construction.
* It's more verbose.
* It's imperative instead of an object literal. Configs as object literals are beautiful.

(The grouping with object literals could have been done with nested object literals, at the cost of some verbosity.)

Anyway, I don't see a way to get the benefits of the previous design _and_ having extensibility. So, I'm not opposed to landing this. It just hurts my eyes, I can live with that. 😆

Reply:
> You can easily forget to call setupBuildConfig. The resulting BuildConfig no longer correct by construction.

The bundling tools now pass a `verifyConfig` function down which will verify the configuration that is passed to invoke hooks is complete and valid ("by construction") - it's bundling tool responsibility because only it knows what extensions are available.

> It's imperative instead of an object literal. Configs as object literals are beautiful.

That's fair. Though the imperative style is necessary due to extensibility. But any code - e.g. the testing code in `package:native_toolchain_c` can always make a non-imperative wrapper around it taking object literals - *for a specific set of extensions*.

The current way of crafting hand-written configs seems also a little too low-level: Instead of crafting a `config.json` and invoking the hook directly, I think it makes more sense to test it the way it's going to run in real life - via `package:native_assets_builder` (either via actual flutter / dart sdk or a testing framework that wraps  `package:native_assets_builder`).

> Anyway, I don't see a way to get the benefits of the previous design and having extensibility. So, I'm not opposed to landing this. It just hurts my eyes, I can live with that. 😆

I can related a bit.